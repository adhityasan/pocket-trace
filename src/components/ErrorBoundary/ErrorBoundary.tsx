/* eslint-disable no-console */
import React, { PureComponent, ErrorInfo, ReactNode } from "react";
import { Text, View } from "react-native";

import { APP_NAME } from "@services/constants/extra";

interface IProps {
  children: ReactNode;
  isConsoleError?: boolean;
}

interface IState {
  hasError: boolean;
}

class ErrorBoundary extends PureComponent<IProps, IState> {
  state = {
    hasError: false
  };

  static getDerivedStateFromError() {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error: Error, info: ErrorInfo) {
    const { isConsoleError = true } = this.props;
    if (isConsoleError) {
      console.groupCollapsed(`ERROR ${APP_NAME} Occured!!!`);
      console.error("[ErrorBoundary] Error message:", error.message);
      console.error("[ErrorBoundary] Error stack:", error.stack);
      console.error("[ErrorBoundary] Component stack:", info.componentStack);
      console.groupEnd();
    }
  }

  render() {
    const { hasError } = this.state;
    const { children } = this.props;

    if (hasError) {
      return (
        <View testID="error-boundary-root">
          <Text style={{ fontSize: 24, fontWeight: "bold", textAlign: "center" }}>
            Oops Something Went Wrong!
          </Text>
        </View>
      );
    }

    return <>{children}</>;
  }
}

export default ErrorBoundary;
