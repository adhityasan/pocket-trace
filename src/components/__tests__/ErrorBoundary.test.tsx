import { Component } from "react";
import { Text } from "react-native";
import { create, act } from "react-test-renderer";

import ErrorBoundary from "@components/ErrorBoundary";

class TriggerError extends Component<any, any> {
  // this components triggers error by rendering undefined state key
  render() {
    return <Text>{this.state.undefinedStateKey}</Text>;
  }
}

describe("components/ErrorBoundary", () => {
  it("renders correctly", () => {
    expect.assertions(2);

    const tree = create(
      <ErrorBoundary isConsoleError={false}>
        <TriggerError />
      </ErrorBoundary>
    ).toJSON();

    act(() => {
      expect(tree).toMatchSnapshot();
      expect(tree).toHaveProperty("props.testID", "error-boundary-root");
    });
  });
});
