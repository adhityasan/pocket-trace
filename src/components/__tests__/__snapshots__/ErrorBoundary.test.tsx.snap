// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ErrorBoundary renders correctly 1`] = `
<View
  testID="error-boundary-root"
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <Text
      style={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Oops Something Went Wrong!
    </Text>
  </RCTSafeAreaView>
</View>
`;
